Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,182 @@
+// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Web;
+using Remotion.Context;
+using Remotion.ServiceLocation;
+using Remotion.Utilities;
+
+namespace Remotion.Web.Context
+{
+#if !NETFRAMEWORK
+  [ImplementationFor(typeof(ISafeContextStorageProvider), Position = 0)]
+  public class HttpContextStorageProvider : ISafeContextStorageProvider, ISafeContextBoundaryStrategy
+  {
+    private struct HashTableAccessor
+    {
+      private class WrappedHashtable
+      {
+        public HttpContext Context { get; }
+        public Hashtable Data { get; }
+
+        public WrappedHashtable (HttpContext context, Hashtable data)
+        {
+          Context = context;
+          Data = data;
+        }
+      }
+
+      private const string c_dataDictionaryKey = "Rm.DD";
+
+      private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable;
+
+      public HashTableAccessor (bool _)
+      {
+        _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
+      }
+
+      public bool TryGetCachedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out Hashtable hashtable)
+      {
+        if (_dataTable.Value!.TryGetTarget(out var target) && target.Context == httpContext)
+        {
+          hashtable = target.Data;
+          return true;
+        }
+
+        var wrappedHashtable = (WrappedHashtable?)httpContext.Items[c_dataDictionaryKey];
+        hashtable = wrappedHashtable?.Data;
+
+        if (wrappedHashtable != null)
+        {
+          _dataTable.Value!.SetTarget(wrappedHashtable);
+          return hashtable != null;
+        }
+
+        return false;
```

Comment:
I think we can assume that `WrappedHashtable.Data` is always non-null? If that is the case I would rewrite this to make it a bit clearer what is happening. As it stands, we switch between `hashtable` and `wrappedHashtable` for null checks when their nullability should always be the same.

```suggestion
        var wrappedHashtable = (WrappedHashtable?)httpContext.Items[c_dataDictionaryKey];
        if (wrappedHashtable != null)
        {
          hashtable = wrappedHashtable?.Data;
          _dataTable.Value!.SetTarget(wrappedHashtable);
          return true;
        }

        hashtable = null;
        return false;
```